kernel/src/syscall/waitid.rs

```diff
@@ -3,28 +3,69 @@
 use super::SyscallReturn;
 use crate::{
     prelude::*,
-    process::{do_wait, ProcessFilter, WaitOptions},
+    process::{
+        do_wait,
+        signal::{c_types::siginfo_t, constants::SIGCHLD},
+        ProcessFilter, WaitOptions,
+    },
 };
 
 pub fn sys_waitid(
     which: u64,
     upid: u64,
-    _infoq_addr: u64,
+    infoq_addr: u64,
     options: u64,
-    _rusage_addr: u64,
     ctx: &Context,
 ) -> Result<SyscallReturn> {
-    // FIXME: what does infoq and rusage use for?
     let process_filter = ProcessFilter::from_which_and_id(which, upid as _)?;
     let wait_options = WaitOptions::from_bits(options as u32)
         .ok_or(Error::with_message(Errno::EINVAL, "invalid options"))?;
 
+    // Check for waitid options
+    if !wait_options
+        .intersects(WaitOptions::WSTOPPED | WaitOptions::WCONTINUED | WaitOptions::WEXITED)
+    {
+        return_errno_with_message!(
+            Errno::EINVAL,
+            "at least one of WSTOPPED, WCONTINUED, or WEXITED should be specified"
+        );
+    }
+
     let wait_status =
         do_wait(process_filter, wait_options, ctx).map_err(|err| match err.error() {
             Errno::EINTR => Error::new(Errno::ERESTARTSYS),
             _ => err,
         })?;
 
-    let pid = wait_status.map_or(0, |wait_status| wait_status.pid());
-    Ok(SyscallReturn::Return(pid as _))
+    let Some(wait_status) = wait_status else {
+        return Ok(SyscallReturn::Return(0));
+    };
+
+    let siginfo = {
+        let mut siginfo = siginfo_t::new(SIGCHLD, wait_status.si_code());
+
+        let pid = wait_status.pid();
+        let uid = wait_status.uid();
+        siginfo.set_pid_uid(pid, uid);
+
+        let status_code = decode_status_code(wait_status.status_code());
+        siginfo.set_status(status_code as i32);
+
+        siginfo
+    };
+
+    ctx.user_space().write_val(infoq_addr as usize, &siginfo)?;
+
+    Ok(SyscallReturn::Return(0))
+}
+
+fn decode_status_code(status_code: u32) -> u32 {
+    const KILL_STATUS_MASK: u32 = 0xff;
+    // If the status code is a kill status, we return the exit code directly.
+    // Otherwise, we shift it right by 8 bits to get the actual exit code.
+    if (status_code & KILL_STATUS_MASK) == 0 {
```

Comment:
The mask is used to determine whether a process has exited normally. When the process exits normally, the lowest 8 bits will be zero. I suggest renaming this mask to `NORMAL_EXIT_MASK`. The comment could be improved as follows:

```rust
// If the process exits normally, the lowest 8 bits of `status_code` will be zero. 
// In this case, we return the actual exit code by shifting the `status_code` right by 8 bits.
// Otherwise, we return the `status_code` as is.

```